File: DeiC HPC Usage Data/DeiC HPC Usage Data/CenterDaily.cs
using System;

namespace DeiC_HPC_Usage_Data
{
    public class CenterDaily
    {
        /*
         * Each HPC center has a unique ID. This is defined as a GUID/UUID
         */
        public Guid HPCCenterId { get; set; }
        /*
         * In case of sub centers they can use a sub id. This is defined as a GUID/UUID
         */
        public Guid SubHPCCenterId { get; set; }
        /*
         * Date for the entry in ISO 8601 format.
         * Please ensure to a correct ISO8601-1:2019 date format.
         * Access time https://en.wikipedia.org/wiki/ISO_8601
         * As this pr day. Time format should just be 00:00:01 etc or just anotner time of the day.
         */
        public DateTime Date { get; set; }
        /*
         * User must have a ORCID. This needs to be collected when loging in.
         * Please use: https://support.orcid.org/hc/en-us/articles/360006897674-Structure-of-the-ORCID-Identifier
         * It is based on https://www.iso.org/standard/44292.html
         */
        public string ORCID { get; set; }

        /*
         * A local id for the user. This must be the same for alle logins from that user.
         * Whe national AAI is ready use the national id. This is the identity provided om the AAI.
         */
        public string LocalId { get; set; }

        /*
         * Each project that are assigned usage time have a generated project id. The format of the ID is string.
         * Change: From GUID to follow the DeiC project format: https://www.deic.dk/en/Supercomputing/Instructions-and-Guides/Remember-to-acknowledge-the-use-of-national-hpc
         */
        public string DEICProjectId { get; set; }
        /*
         * Each university is defined as a constant. New will be added if needed.
         */
        public UniversityId UniversityId { get; set; }
        /*
         * In case of unknown, industry or other is used please specify in the IdExpanded field.
         */
        public string IdExpanded { get; set; }
        /*
         * Each access type is defined as a constand.
         * The access type can be read from the project id also. It is part of basic data type to enable better filtering
         * See AccessType datatype.
         */
        public AccessType AccessType { get; set; }
        /*
         * Max CPU core time in hours
         */
        public ulong MaxCPUCoreTime { get; set; }
        /*
         * Used CPU core time in hours
         */
        public ulong UsedCPUCoretime { get; set; }
        /*
         * Max GPU core time in hours
         */
        public ulong MaxGPUCoreTime { get; set; }
        /*
         * Used GPU core time in hours
         */
        public ulong UsedGPUCoretime { get; set; }
        /*
         * Storage space in MB
         */
        public ulong StorageUsedInMB { get; set; }
        /*
        * Network usage in MB
        */
        public ulong? NetworkUsageInMB { get; set; }
        /*
         * Network avg in Mbps
         */
        public double? NetworkAvgUsage { get; set; }
        /*
         * Max node time. For Type 4 only as they do not have CPU/GPU core times.
         */
        public ulong? MaxNodeTime { get; set; }
        /*
         * Used node time. For Type 4 only as they do not have CPU/GPU core times.
         */
        public ulong? UsedNoteTime { get; set; }
    }

}

File: DeiC HPC Usage Data/DeiC HPC Usage Data/Python/CenterDaily.py
# This code parses date/times, so please
#
#     pip install python-dateutil
#
# To use this code, make sure you
#
#     import json
#
# and then, to convert JSON from a string, do
#
#     result = center_daily_from_dict(json.loads(json_string))

from dataclasses import dataclass
from uuid import UUID
from datetime import datetime
from typing import Any, List, TypeVar, Callable, Type, cast
import dateutil.parser


T = TypeVar("T")


def from_datetime(x: Any) -> datetime:
    return dateutil.parser.parse(x)


def from_str(x: Any) -> str:
    assert isinstance(x, str)
    return x


def from_int(x: Any) -> int:
    assert isinstance(x, int) and not isinstance(x, bool)
    return x


def from_list(f: Callable[[Any], T], x: Any) -> List[T]:
    assert isinstance(x, list)
    return [f(y) for y in x]


def to_class(c: Type[T], x: Any) -> dict:
    assert isinstance(x, c)
    return cast(Any, x).to_dict()


@dataclass
class CenterDailyElement:
    hpc_center_id: UUID
    sub_hpc_center_id: UUID
    date: datetime
    orcid: str
    deic_project_id: str
    university_id: int
    id_expanded: str
    access_type: int
    max_cpu_core_time: int
    used_cpu_coretime: int
    max_gpu_core_time: int
    used_gpu_coretime: int
    storage_used_in_gb: int
    network_usage_in_gb: int
    network_avg_usage: int
    max_node_time: int
    used_note_time: int

    @staticmethod
    def from_dict(obj: Any) -> 'CenterDailyElement':
        assert isinstance(obj, dict)
        hpc_center_id = UUID(obj.get("hpcCenterId"))
        sub_hpc_center_id = UUID(obj.get("subHPCCenterId"))
        date = from_datetime(obj.get("date"))
        orcid = from_str(obj.get("orcid"))
        deic_project_id = from_str(obj.get("deicProjectId"))
        university_id = from_int(obj.get("universityId"))
        id_expanded = from_str(obj.get("idExpanded"))
        access_type = from_int(obj.get("accessType"))
        max_cpu_core_time = from_int(obj.get("maxCPUCoreTime"))
        used_cpu_coretime = from_int(obj.get("usedCPUCoretime"))
        max_gpu_core_time = from_int(obj.get("maxGPUCoreTime"))
        used_gpu_coretime = from_int(obj.get("usedGPUCoretime"))
        storage_used_in_gb = from_int(obj.get("storageUsedInGB"))
        network_usage_in_gb = from_int(obj.get("networkUsageInGB"))
        network_avg_usage = from_int(obj.get("networkAvgUsage"))
        max_node_time = from_int(obj.get("maxNodeTime"))
        used_note_time = from_int(obj.get("usedNoteTime"))
        return CenterDailyElement(hpc_center_id, sub_hpc_center_id, date, orcid, deic_project_id, university_id, id_expanded, access_type, max_cpu_core_time, used_cpu_coretime, max_gpu_core_time, used_gpu_coretime, storage_used_in_gb, network_usage_in_gb, network_avg_usage, max_node_time, used_note_time)

    def to_dict(self) -> dict:
        result: dict = {}
        result["hpcCenterId"] = str(self.hpc_center_id)
        result["subHPCCenterId"] = str(self.sub_hpc_center_id)
        result["date"] = self.date.isoformat()
        result["orcid"] = from_str(self.orcid)
        result["deicProjectId"] = from_str(self.deic_project_id)
        result["universityId"] = from_int(self.university_id)
        result["idExpanded"] = from_str(self.id_expanded)
        result["accessType"] = from_int(self.access_type)
        result["maxCPUCoreTime"] = from_int(self.max_cpu_core_time)
        result["usedCPUCoretime"] = from_int(self.used_cpu_coretime)
        result["maxGPUCoreTime"] = from_int(self.max_gpu_core_time)
        result["usedGPUCoretime"] = from_int(self.used_gpu_coretime)
        result["storageUsedInGB"] = from_int(self.storage_used_in_gb)
        result["networkUsageInGB"] = from_int(self.network_usage_in_gb)
        result["networkAvgUsage"] = from_int(self.network_avg_usage)
        result["maxNodeTime"] = from_int(self.max_node_time)
        result["usedNoteTime"] = from_int(self.used_note_time)
        return result


def center_daily_from_dict(s: Any) -> List[CenterDailyElement]:
    return from_list(CenterDailyElement.from_dict, s)


def center_daily_to_dict(x: List[CenterDailyElement]) -> Any:
    return from_list(lambda x: to_class(CenterDailyElement, x), x)
